src/PimlicoERC20Paymaster.sol

```diff
@@ -127,38 +120,56 @@ contract PimlicoERC20Paymaster is BasePaymaster {
     /// @param mode The post-operation mode (either successful or reverted).
     /// @param context The context containing the token amount and user sender address.
     /// @param actualGasCost The actual gas cost of the transaction.
-    function _postOp(PostOpMode mode, bytes calldata context, uint256 actualGasCost) internal override {
-        if (mode == PostOpMode.postOpReverted) {
-            return; // Do nothing here to not revert the whole bundle and harm reputation
-        }
-        unchecked {
-            uint192 tokenPrice = fetchPrice(tokenOracle);
-            uint192 nativeAsset = fetchPrice(nativeAssetOracle);
-            uint256 cachedPrice = previousPrice;
-            uint192 price = nativeAsset * uint192(tokenDecimals) / tokenPrice;
-            uint256 cachedUpdateThreshold = priceUpdateThreshold;
-            if (
-                uint256(price) * priceDenominator / cachedPrice > priceDenominator + cachedUpdateThreshold
-                    || uint256(price) * priceDenominator / cachedPrice < priceDenominator - cachedUpdateThreshold
-            ) {
-                previousPrice = uint192(int192(price));
-                cachedPrice = uint192(int192(price));
+    function _postOp(PostOpMode mode, bytes calldata context, uint256 actualGasCost, uint256 actualUserOpFeePerGas)
+        internal
+        override
+    {
+        uint256 prefundTokenAmount = uint256(bytes32(context[0:32]));
+        uint192 tokenPrice = uint192(bytes24(context[32:56]));
+        address sender = address(bytes20(context[56:76]));
+        uint256 actualTokenNeeded = (actualGasCost + REFUND_POSTOP_COST * actualUserOpFeePerGas) * priceMarkup
+            * tokenPrice / (1e18 * priceDenominator); // We use tx.gasprice here since we don't know the actual gas price used by the user
+
+        // If the initially provided token amount is greater than the actual amount needed, refund the difference
+        if (prefundTokenAmount > actualTokenNeeded) {
+            if (context.length == 96) {
+                address payer = address(bytes20(context[76:96]));
+
+                bytes memory innerCall =
+                    abi.encodeCall(this.safeTransferFrom, (token, sender, address(this), actualTokenNeeded));
+
+                bool success;
+                // solhint-disable-next-line no-inline-assembly
+                assembly ("memory-safe") {
+                    success := call(gas(), address(), 0, add(innerCall, 0x20), mload(innerCall), 0, 32)
+                }
+
```

Comment:
I would move the lib code into an internal function that returns success/failure